### PR TITLE
Don't use numeric IDs for internal PHP serialization.

### PR DESCRIPTION
**Pull request by brightbyte**
This changes Snak serialization to use the full string representation of property IDs.

NOTE: serialize/unserialize is used mainly when cloning Statements.
NOTE: it's unclear if this completely fixes T157442
NOTE: this changes snak hashes, and consequently reference hashes! This may break reference update/removal!

[Bug: T157442](https://phabricator.wikimedia.org/T157442)

### CODE REVIEW

**Comment by brightbyte** [comment]
pinging @manicki @thiemowmde

File: src/Entity/EntityIdValue.php [hunk 1]
```diff
@@ -31,8 +31,8 @@ public function __construct( EntityId $entityId ) {
 	 */
 	public function serialize() {
 		return json_encode( [
-			$this->entityId->getEntityType(),
-			$this->getNumericId()
+			get_class( $this->entityId ),
```
Comment (JeroenDeDauw) [inline src/Entity/EntityIdValue.php]: Why this change? Kinda scary to do `new $userInput()`

Comment (brightbyte) [inline src/Entity/EntityIdValue.php]: Yes, it's not pretty. But all we use this for is cloning statements.

The change is here because the old code assumes two things: a) the ID can be represented as a number (it can't) and b) we have a well known set of entity types (we don't). 

So, having the choice to try and access some kind of entity id type registry via a global, or putting the class name into the serialization, I opted for the latter.

File: src/Entity/EntityIdValue.php [hunk 2]
```diff
@@ -58,10 +58,14 @@ private function getNumericId() {
 	 * @throws IllegalValueException
 	 */
 	public function unserialize( $serialized ) {
-		list( $entityType, $numericId ) = json_decode( $serialized );
+		list( $type, $id ) = json_decode( $serialized );
 
 		try {
-			$entityId = LegacyIdInterpreter::newIdFromTypeAndNumber( $entityType, $numericId );
+			if ( is_string( $id ) ) {
```
Comment (thiemowmde) [inline src/Entity/EntityIdValue.php]: This needs at least a `class_exists` check and a fallback solution (or return `null`?).

File: src/Snak/SnakObject.php [hunk 3]
```diff
@@ -113,7 +113,24 @@ public function serialize() {
...
+	 * @param mixed $unserialized
+	 *
+	 * @return PropertyId
+	 */
+	protected static function newPropertyId( $unserialized ) {
+		if ( is_int( $unserialized ) ) {
+			return PropertyId::newFromNumber( $unserialized );
+		} elseif ( is_string( $unserialized ) ) {
+			return new PropertyId( $unserialized );
+		} elseif ( $unserialized instanceof PropertyId ) {
```
Comment (thiemowmde) [inline src/Snak/SnakObject.php]: Which code path needs this?

File: src/Entity/EntityIdValue.php (hunk 1 above)
Comment (JeroenDeDauw) [inline src/Entity/EntityIdValue.php]: My remark is not about prettiness, it is about security.

> But all we use this for is cloning statements.

That does no reassure me. Kinda like saying "no need to escape this SQL argument since right now only we call this function with non-user-input".

> So, having the choice to try and access some kind of entity id type registry via a global, or putting the class name into the serialization, I opted for the latter.

Why not just (de)serialize the ID?

Comment (brightbyte) [inline src/Entity/EntityIdValue.php]: Oh, I definitely do not want this in the database, or in our external representation! As far as I am aware, this is not used in any way with our JSON encoding. We only use the php-serialization temporarily, to do deep clones. PHP-Serialization is extremely brittle (sensitive against changes to private members, etc), so it should never be used for persistence.

The alternative would be to change the way we do cloning, and avoid copying immutable objects when we clone. That would be nice, but I can't think of a reliable and easy way of doing this. serialize/unserialize is a very convenient method to do deep cloning safely. But it's somewhat wasteful, since it also copies immutable objects.

File: src/Snak/SnakObject.php (hunk 3 above)
Comment (brightbyte) [inline src/Snak/SnakObject.php]: Probably none atm, it's here for completeness. I don't have strong feelings about having it.

File: src/Entity/EntityIdValue.php (hunk 1 above)
Comment (brightbyte) [inline src/Entity/EntityIdValue.php]: We can't deserialize the ID without access to a service. How do we get the right service instance? This is particularly tricky for entity parsing, since we use different EntityIdParsers for data coming from different repos.

Comment (brightbyte) [inline src/Entity/EntityIdValue.php]: But I appreciate your concern for security. I'm not very happy about this solution. I just don't see an alternative, apart from not using serialize/unserialize for cloning statements.

Comment (JeroenDeDauw) [inline src/Entity/EntityIdValue.php]: > We can't deserialize the ID without access to a service.

Huh?

```php
deserialize( $entityId );
```

Comment (thiemowmde) [inline src/Entity/EntityIdValue.php]: It's correct that these `serialize` formats are not used in our JSON encoding, but they are used to build summaries.

Comment (thiemowmde) [inline src/Entity/EntityIdValue.php]: I uploaded a draft utilizing this approach as #718.

Comment (manicki) [inline src/Entity/EntityIdValue.php]: For what is worth: if we didn't want to use PHP serialization for EntityIdValue that @thiemowmde is suggesting in https://github.com/wmde/WikibaseDataModel/pull/718 (I am not sure what are implications with that approach, if this serialization is never stored) another option (theoretically) could be using `EntityIdComposer` which is meant as a entity-type-agnostic (kinda) replacement for `LegacyIdInterpreter`. The obvious problem is how to pull `EntityIdComposer` in here, so it makes sense (it is now bound to Wikibase git repo and entity type definitions out there). I cannot think of good way to achieve this, so this comment is mostly noise.

Also theoretically, another option would be to use value of EntityId::getSerialization in EntityIdValue::serialize, and use EntityIdParser in EntityIdValue::unserialize but that seems so wrong.

Comment (brightbyte) [inline src/Entity/EntityIdValue.php]: @JeroenDeDauw Yes, I realized this morning that I misunderstood. deserialize( $this->entityId ) will work, sure.

File: src/Entity/EntityIdValue.php (hunk 2 above)
Comment (brightbyte) [inline src/Entity/EntityIdValue.php]: If the class doesn't exist, this should fail hard. Ok, we could throw an exception instead of a fatal error.

**Comment by brightbyte** [comment]
I talked to Thiemo about this. We decided to use his patch #718 for EntityIdValue and the EntityId classes, and my changes for SnakObject and PropertyValueSnak. I will remove the changes to EntityIdValue from this pull request to avoid conflicts. The two pull requests do not depend on each other, but are conceptually related, and should become part of the same release. 

Note that this is a breaking change only if we consider the php serialization to be a stable external interface. I don't think we have an explicite statement about this anywhere.

File: src/Entity/EntityIdValue.php (hunk 1 above)
Comment (brightbyte) [inline src/Entity/EntityIdValue.php]: @thiemowmde you wrote:
>  they are used to build summaries.

when/where?  I can't find it.

**Comment by brightbyte** [comment]
NOTE: this changes snak hashes, and consequently reference hashes! This may break reference update/removal!

File: src/Snak/SnakObject.php (hunk 3 above)
Comment (brightbyte) [inline src/Snak/SnakObject.php]: Still want me to remove this? This is currently blocking the pull  request...

File: src/Snak/SnakObject.php [hunk 4]
```diff
@@ -102,7 +102,7 @@ public function equals( $target ) {
 	 * @return string
 	 */
 	public function serialize() {
-		return serialize( $this->propertyId->getNumericId() );
+		return serialize( $this->propertyId->getSerialization() );
```
Comment (thiemowmde) [inline src/Snak/SnakObject.php]: This is a bit pointless. Please replace with a straight `return $this->propertyId->getSerialization();`.

File: src/Snak/SnakObject.php [hunk 5]
```diff
@@ -113,7 +113,22 @@ public function serialize() {
...
+		$this->propertyId = self::newPropertyId( unserialize( $serialized ) );
+	}
+
+	/**
+	 * @param mixed $unserialized
+	 *
+	 * @return PropertyId
+	 */
+	protected static function newPropertyId( $unserialized ) {
+		if ( is_int( $unserialized ) ) {
```
Comment (thiemowmde) [inline src/Snak/SnakObject.php]: This also needs redoing with the change above, and should probably be inlined above:
```php
$numericId = unserialize( $serialized );
$this->propertyId = is_int( $numericId )
    ? PropertyId::newFromNumber( $numericId )
    : new PropertyId( $serialized );
```

**Comment by brightbyte** [comment]
NOTE: once this is merged, we should make a release.

File: src/Snak/SnakObject.php [hunk 6]
```diff
@@ -113,7 +113,34 @@ public function serialize() {
 	 * @param string $serialized
 	 */
 	public function unserialize( $serialized ) {
-		$this->propertyId = PropertyId::newFromNumber( unserialize( $serialized ) );
+		$this->propertyId = self::newPropertyId( $serialized );
+	}
+
+	/**
+	 * @param string $serialized
```
Comment (thiemowmde) [inline src/Snak/SnakObject.php]: string|int

File: src/Snak/PropertyValueSnak.php [hunk 7]
```diff
@@ -64,8 +64,8 @@ public function serialize() {
 	 * @param string $serialized
 	 */
 	public function unserialize( $serialized ) {
-		list( $numericId, $dataValue ) = unserialize( $serialized );
-		$this->__construct( $numericId, $dataValue );
+		list( $propertyId, $dataValue ) = unserialize( $serialized );
+		$this->__construct( self::newPropertyId( $propertyId ), $dataValue );
```
Comment (thiemowmde) [inline src/Snak/PropertyValueSnak.php]: Mark the date in your calendar. ;-) This might be the first time I don't like the fact that this relies on a protected method from a base class. Can be much simpler:
```php
$this->__construct( is_int( $propertyId ) ? $propertyId : new PropertyId( $propertyId ), $dataValue );
```

File: src/Snak/SnakObject.php [hunk 8]
```diff
@@ -113,7 +113,34 @@ public function serialize() {
...
+
+		try {
+			// full ID, as a string, not serialized
+			return new PropertyId( $serialized );
+		} catch ( InvalidArgumentException $ex ) {
+			// noop
+		}
+
+		$unserialized = unserialize( $serialized );
+		if ( is_int( $unserialized ) ) {
```
Comment (thiemowmde) [inline src/Snak/SnakObject.php]: Sorry, but this code is weird. It does an `is_int` check twice? The reason is that this method is used for two very different use cases:
* In Value snaks this can be called with either `1` or `"P1"`.
* In NoValue and SomeValue snaks this can be called with `"i:42;"` or `"P1"`.

The problem is: there is no need to make these two use cases compatible with each other.

**Comment by brightbyte** [comment]
Thanks @thiemowmde  for refactoring the PropertyId instantiation! +1 for that.

**Comment by JeroenDeDauw** [comment]
+1

File: tests/unit/Snak/PropertyValueSnakTest.php [hunk 9]
```diff
@@ -121,15 +121,62 @@ public function notEqualsProvider() {
...
-	public function testUnserialize() {
+	public function provideDataToUnserialize() {
+		$p2 = new PropertyId( 'P2' );
+		$p2foo = new PropertyId( 'foo:P2' );
+		$value = new StringValue( 'b' );
+
+		return [
+			'legacy' => [
+				new PropertyValueSnak( $p2, $value ),
+				'a:2:{i:0;i:2;i:1;C:22:"DataValues\StringValue":1:{b}}'
```
Comment (manicki) [inline tests/unit/Snak/PropertyValueSnakTest.php]: It does not affect test result but I believe to be 100% accurate "i:2" here should become "i:1". Minor thing, I am going to fix this, and squash into other commits.

File: src/Snak/SnakObject.php [hunk 10]
```diff
@@ -97,12 +97,12 @@ public function equals( $target ) {
 	/**
 	 * @see Serializable::serialize
 	 *
-	 * @since 0.1
+	 * @since 7.0
```
Comment (manicki) [inline src/Snak/SnakObject.php]: I wonder is this correct to change this number here. Change here is indeed breaking change but the interface of the methods does not change. So I guess version number change is to show that serialization format has changed in the release 7.0. I am find with that I just wonder if this is "how it should be done" (tm).

File: tests/unit/Snak/PropertyValueSnakTest.php (hunk 9 above)
Comment (manicki) [inline tests/unit/Snak/PropertyValueSnakTest.php]: Please ignore this comment. I looked at wrong spot in the old code. All good here.

File: src/Snak/SnakObject.php (hunk 10 above)
Comment (thiemowmde) [inline src/Snak/SnakObject.php]: I believe the returned value is a crucial part of the contract of a method. If the format of the return value changes, it's a different method. The idea of the since tag is to make this obvious.

Comment (JeroenDeDauw) [inline src/Snak/SnakObject.php]: We have some places where we bumped the number but then mention the method was already there in some other form or something along those lines

Comment (manicki) [inline src/Snak/SnakObject.php]: Yea, something like this seems reaonable.
